processor/src/system/mod.rs

```diff
@@ -146,6 +149,13 @@ impl System {
         self.fn_hash_trace[1][clk] = self.fn_hash[1];
         self.fn_hash_trace[2][clk] = self.fn_hash[2];
         self.fn_hash_trace[3][clk] = self.fn_hash[3];
+
+        // Check that maximum number of cycles is not exceeded.
+        if self.clk() > max_cycles {
+            return Err(ExecutionError::CycleLimitExceeded(max_cycles));
+        }
```

Comment:
minor nit: I'd probably move this up to be right after line 139 and would probably just use `self.clk` (rather than `self.clk()`) for consistency.